engine/src/snapshot_api/rocks_snapshot.rs

```diff
@@ -340,39 +330,43 @@ impl SnapshotApi for RocksSnapshot {
     }
 
     #[inline]
-    async fn read_exact(&mut self, mut buf: &mut [u8]) -> std::io::Result<()> {
-        while !buf.is_empty() {
-            // the return value of read function is not larger than the input buf's length.
-            #[allow(clippy::indexing_slicing)]
+    async fn read_exact(&mut self, buf: &mut BytesMut) -> std::io::Result<()> {
+        let origin_cap = buf.capacity();
+        while buf.len() < buf.capacity() {
+            // the return value of read function is not larger than the input buf's capacity.
             match self.read(buf).await {
-                Ok(0) => break,
                 Ok(n) => {
-                    let tmp = buf;
-                    buf = &mut tmp[n..];
+                    if 0 == n {
+                        break;
+                    }
                 }
                 Err(ref e) if e.kind() == ErrorKind::Interrupted => {}
                 Err(e) => return Err(e),
             }
         }
-        if buf.is_empty() {
+        if buf.len() == buf.capacity() {
             Ok(())
         } else {
             Err(std::io::Error::new(
                 ErrorKind::UnexpectedEof,
-                "failed to fill whole buffer",
+                format!(
+                    "failed to fill whole buffer: {} {} {}",
+                    buf.len(),
+                    buf.capacity(),
+                    origin_cap,
+                ),
             ))
         }
     }
 
     #[inline]
-    async fn write_all(&mut self, mut buf: &[u8]) -> std::io::Result<()> {
+    async fn write_all(&mut self, mut buf: Bytes) -> std::io::Result<()> {
         let mut written_size = 0;
         // the return value of write function is not larger than the input buf's length.
         #[allow(clippy::indexing_slicing)]
         while !buf.is_empty() {
```

Comment:
You can use `has_remaining` to replace `buf.is_empty` and we don't need to maintain a `written_size` variable to track how many bytes we read.